src/hybridmap/HybridNeutralAtomMapper.cpp

```diff
@@ -0,0 +1,1293 @@
+    }
+    // then do all shuttling based mapping gates
+    while (!this->frontLayerShuttling.empty()) {
+      GateList gatesToExecute;
+      while (gatesToExecute.empty()) {
+        ++i;
+        if (this->verbose) {
+          std::cout << "iteration " << i << '\n';
+        }
+        auto bestMove = findBestAtomMove();
+        updateMappingMove(bestMove);
+        gatesToExecute = getExecutableGates(frontLayer.getGates());
+      }
+      mapAllPossibleGates(frontLayer);
+      lookaheadLayer.initLayerOffset(frontLayer.getIteratorOffset());
+      reassignGatesToLayers(frontLayer.getGates(), lookaheadLayer.getGates());
+      if (this->verbose) {
+        printLayers();
+      }
+    }
+  }
+  if (this->verbose) {
+    std::cout << "nSwaps: " << nSwaps << '\n';
+    std::cout << "nMoves: " << nMoves << '\n';
+  }
+  return mappedQc;
+}
+
+void NeutralAtomMapper::mapAllPossibleGates(qc::NeutralAtomLayer& layer) {
+  // map single qubit gates
+  for (const auto* opPointer : layer.getMappedSingleQubitGates()) {
+    mapGate(opPointer);
+  }
+  layer.removeGatesAndUpdate({});
+  // check and map multi qubit gates
+  auto executableGates = getExecutableGates(layer.getGates());
+  while (!executableGates.empty()) {
+    for (const auto* opPointer : layer.getMappedSingleQubitGates()) {
+      mapGate(opPointer);
+    }
+    for (const auto* opPointer : executableGates) {
+      mapGate(opPointer);
+    }
+    layer.removeGatesAndUpdate(executableGates);
+    executableGates = getExecutableGates(layer.getGates());
+  }
+}
+
+QuantumComputation NeutralAtomMapper::convertToAod(qc::QuantumComputation& qc) {
+  // decompose SWAP gates
+  CircuitOptimizer::decomposeSWAP(qc, false);
+  CircuitOptimizer::replaceMCXWithMCZ(qc);
+  CircuitOptimizer::singleQubitGateFusion(qc);
+  CircuitOptimizer::flattenOperations(qc);
+  // decompose AOD moves
+  AodScheduler aodScheduler(this->arch);
+  mappedQcAOD = aodScheduler.schedule(qc);
+  if (this->verbose) {
+    std::cout << "nMoveGroups: " << aodScheduler.getNMoveGroups() << '\n';
+  }
+  return mappedQcAOD;
+}
+
+void NeutralAtomMapper::reassignGatesToLayers(const GateList& frontGates,
+                                              const GateList& lookaheadGates) {
+  // assign gates to gates or shuttling
+  this->frontLayerGate.clear();
+  this->frontLayerShuttling.clear();
+  for (const auto& gate : frontGates) {
+    if (swapGateBetter(gate)) {
+      this->frontLayerGate.push_back(gate);
+    } else {
+      this->frontLayerShuttling.push_back(gate);
+    }
+  }
+
+  this->lookaheadLayerGate.clear();
+  this->lookaheadLayerShuttling.clear();
+  for (const auto& gate : lookaheadGates) {
+    if (swapGateBetter(gate)) {
+      this->lookaheadLayerGate.push_back(gate);
+    } else {
+      this->lookaheadLayerShuttling.push_back(gate);
+    }
+  }
+}
+
+void qc::NeutralAtomMapper::mapGate(const Operation* op) {
+  if (op->getType() == qc::OpType::I) {
+    return;
+  }
+  // Safety check
+  // TODO remove
```

Comment:
Can this todo be completed?

Reply:
The check prevent commuting gates to be applied multiple times. The TODO was because I thought it also might work without it, but as a result I think this is the best solution. Therefore I just removed the TODO